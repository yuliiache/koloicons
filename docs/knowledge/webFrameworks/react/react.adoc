== React
:toc:

=== 1. https://legacy.reactjs.org/docs/components-and-props.html#function-and-class-components[React component definition (functional, class)]
=== 2. https://react.dev/learn/passing-props-to-a-component[React component props]
- https://legacy.reactjs.org/docs/render-props.html[render props]

=== 3. https://react.dev/learn/state-a-components-memory[React component state]
- initial state
- setState
- stateful vs stateless components

=== 4. React class component lifecycle methods
=== 5.* https://react.dev/reference/react[React hooks API]:
- `useEffect (https://react.dev/learn/removing-effect-dependencies[problem])`, `useState`, `useReducer`, `useMemo`, `useRef`
- https://reactjs.org/docs/hooks-rules.html[rules]
- https://react.dev/learn/reusing-logic-with-custom-hooks[custom hooks]

=== 6. Components composition
- props.children
- composition vs inheritance
- lifting state up
- practical decomposition task

=== 7. https://react.dev/learn/writing-markup-with-jsx[JSX]
=== 8. https://react.dev/reference/react-dom[React DOM]
=== 9. https://react.dev/reference/react-dom/components/common#reference[React DOM Events]
`SyntheticEvent`, `persist event`

=== 10. https://react.dev/reference/react-dom/components#form-components:~:text=and%20dangerouslySetInnerHTML.-,Form%20components,-These%20built%2Din[Forms]
`controlled`,`uncontrolled components`

=== 11. https://legacy.reactjs.org/docs/faq-internals.html#what-is-the-virtual-dom[Virtual DOM]
- https://ru.reactjs.org/docs/reconciliation.html[reconciliation]
- https://github.com/acdlite/react-fiber-architecture[fiber]

=== 12. React code reuse patterns
- https://legacy.reactjs.org/docs/higher-order-components.html[higher-order components]

=== 13. https://react.dev/reference/react/createContext[Context]
`when to use`, `Context.Provider`, `Context.Consumer`, `Context vs Redux`

=== 14. https://react.dev/reference/react-dom/createPortal[Portals]
=== 15.  https://react.dev/reference/react/Component#catching-rendering-errors-with-an-error-boundary[Error Boundaries]

=== 16.  React component lifecycle phases (render, pre-commit, commit)
- https://vmarchesin.medium.com/the-react-lifecycle-step-by-step-47c0db0bfe73#:%7E:text=Pre%2Dcommit%20phase%3A%20Is%20a,any%20side%20effects%20are%20triggered[class components]
- https://react.dev/learn/render-and-commit[fn components]
- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#render-and-commit-phases[...]

=== 19. https://react.dev/reference/react-dom/server[Server rendering]
`renderToPipeableStream`
`renderToStaticNodeStream`
`renderToReadableStream`
`renderToString`
`renderToNodeStream`


=== 20. https://react.dev/reference/react/memo[React.memo], https://react.dev/reference/react/lazy[React.lazy, Suspense p1] https://betterprogramming.pub/routing-and-code-splitting-in-react-fc7315f0bde4[p2]

=== 21. Optimizing Performance
- https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key[keys]
- https://react.dev/reference/react/Profiler[Profiler],
- https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en[Dev Tools],
- https://create-react-app.dev/docs/production-build/[prod build, code splitting],
- https://levelup.gitconnected.com/how-to-render-your-lists-faster-with-react-virtualization-5e327588c910[virtualization]

=== 22. Security
-  https://medium.com/javascript-security/avoiding-xss-in-react-is-still-hard-d2b5c7ad9412[XSS on React Props]
- https://blog.logrocket.com/using-dangerouslysetinnerhtml-in-a-react-application/[dangerouslySetInnerHTML]







