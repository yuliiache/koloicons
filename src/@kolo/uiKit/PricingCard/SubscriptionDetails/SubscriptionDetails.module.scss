@import "@kolo/design";

.advantages {
  display: flex;
  flex-direction: column;
  font-size: 14px;
  font-weight: 500;
  margin-top: 24px;
}

.advantagesItem {
  display: flex;
  flex-direction: column;
  margin-top: 0;
  margin-bottom: 5px;
  line-height: 1.7;
}

.itemTitle {
  color: $lighthGrey;
  font-weight: 400;

  &:first-letter {
    text-transform: uppercase;
  }
}

.itemVal {
  font-weight: 500;
  color: $mainFontColor;

  &:first-letter {
    text-transform: uppercase;
  }
}
