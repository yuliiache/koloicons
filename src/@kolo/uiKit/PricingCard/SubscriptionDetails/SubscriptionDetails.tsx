import {FC} from 'react';

import OptionList from '../OptionList/OptionList';
import styles from './SubscriptionDetails.module.scss';

interface SubscriptionDetailsProps {
  formats: string[];
  attribution: string;
  allowedOptions: string[];
  disallowedOptions: string[];
}

const SubscriptionDetails: FC<SubscriptionDetailsProps> = ({
  formats,
  attribution,
  allowedOptions,
  disallowedOptions,
}) => {
  const formattedFormats = formats && formats.map((format) => format).join(', ');

  return (
    <>
      <div className={styles.advantages}>
        <p className={styles.advantagesItem}>
          <span className={styles.itemTitle}>Formats:</span>
          <span className={styles.itemVal}>{formattedFormats}</span>
        </p>
        <p className={styles.advantagesItem}>
          <span className={styles.itemTitle}>Attribution:</span>
          <span className={styles.itemVal}>{attribution}</span>
        </p>
      </div>
      <dl className={styles.advantages}>
        <OptionList
          title="You can use icons for:"
          list={allowedOptions}
        />
        <OptionList
          title="You can not use icons for:"
          list={disallowedOptions}
        />
      </dl>
    </>
  );
};

export default SubscriptionDetails;
