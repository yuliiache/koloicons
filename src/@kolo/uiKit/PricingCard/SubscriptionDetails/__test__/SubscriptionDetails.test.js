import {cleanup, render} from '@testing-library/react';

import '@testing-library/jest-dom/extend-expect';

import SubscriptionDetails from '../SubscriptionDetails';

describe('SubscriptionDetails Component', () => {
  afterEach(cleanup);

  const commonProps = {
    allowedOptions: ['Option 1', 'Option 2', 'Option 3'],
    disallowedOptions: ['Option 4', 'Option 5', 'Option 6'],
  };

  it('renders snapshot', () => {
    const {asFragment} = render(<SubscriptionDetails {...commonProps} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
