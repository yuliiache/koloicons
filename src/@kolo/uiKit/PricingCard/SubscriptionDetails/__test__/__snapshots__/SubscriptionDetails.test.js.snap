// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubscriptionDetails Component renders snapshot 1`] = `
<DocumentFragment>
  <div
    class="advantages"
  >
    <p
      class="advantagesItem"
    >
      <span
        class="itemTitle"
      >
        Formats:
      </span>
      <span
        class="itemVal"
      />
    </p>
    <p
      class="advantagesItem"
    >
      <span
        class="itemTitle"
      >
        Attribution:
      </span>
      <span
        class="itemVal"
      />
    </p>
  </div>
  <dl
    class="advantages"
  >
    <dt
      class="optionTitle"
    >
      You can use icons for:
    </dt>
    <dd
      class="optionVal"
    >
      <span>
         - 
      </span>
      Option 1
    </dd>
    <dd
      class="optionVal"
    >
      <span>
         - 
      </span>
      Option 2
    </dd>
    <dd
      class="optionVal"
    >
      <span>
         - 
      </span>
      Option 3
    </dd>
    <dt
      class="optionTitle"
    >
      You can not use icons for:
    </dt>
    <dd
      class="optionVal"
    >
      <span>
         - 
      </span>
      Option 4
    </dd>
    <dd
      class="optionVal"
    >
      <span>
         - 
      </span>
      Option 5
    </dd>
    <dd
      class="optionVal"
    >
      <span>
         - 
      </span>
      Option 6
    </dd>
  </dl>
</DocumentFragment>
`;
